x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java

```diff
@@ -409,17 +402,33 @@ private <T extends TransportResponse> void sendWithCrossClusterAccessHeaders(
                             if (roleDescriptorsIntersection.isEmpty()) {
                                 throw authzService.remoteActionDenied(authentication, action, remoteClusterAlias);
                             }
-                            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-                                new CrossClusterAccessHeaders(
-                                    remoteClusterCredentials.credentials(),
-                                    new CrossClusterAccessSubjectInfo(authentication, roleDescriptorsIntersection)
-                                ).writeToContext(threadContext);
-                                sender.sendRequest(connection, action, request, options, contextRestoreHandler);
-                            }
-                        }, e -> contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
+                            final var crossClusterAccessHeaders = new CrossClusterAccessHeaders(
+                                remoteClusterCredentials.credentials(),
+                                new CrossClusterAccessSubjectInfo(authentication, roleDescriptorsIntersection)
+                            );
+                            sendWithCrossClusterAccessHeaders(crossClusterAccessHeaders, connection, action, request, options, handler);
+                        }, e -> handler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
```

Comment:
The error handling here now uses `handler` instead of `contextRestoreHandler`. It means the context won't be restored if there anything goes wrong with `getRoleDescriptorsIntersectionForRemoteCluster`. It may not be important at all. But at the same time, I am not entirely sure either since threadContext leaking can be tricky to trace. So my preference would be keep the previous behaviour. What do you think?

Reply:
I think this is safe since we do not modify the thread context before calling `getRoleDescriptorsIntersectionForRemoteCluster`, `getRoleDescriptorsIntersectionForRemoteCluster` itself uses `wrapPreservingContext` internally, and nothing inside the `Action.wrap() {}` lambda before the `sendWithCrossClusterAccessHeaders` call modifies the thread context. 

I can revert to the previous behavior but I think the current flow reads better and keeps the `ContextRestoreResponseHandler` instantiation closer to the block where we actually modify the context, which also seems preferable. 

Let me know if you feel strongly though or if I'm missing something here -- thread context forever holds a hint of mystery to me 😅